scripts/convert_data.py

```diff
@@ -0,0 +1,243 @@
+    """
+
+    def join_simple_chunk(chunk: str) -> list:
+        if len(chunk.strip()) == 0:
+            return []
+        # keep the newlines, but discard the non-newline whitespace
+        tokens = re.split(r'(\n)|\s', chunk.strip())
+        # the re will return None for splits which were not caught in a group
+        tokens = [x for x in tokens if x is not None]
+        return [token + ' O' if len(token.strip()) > 0 else token for token in tokens]
+
+    converters = {'beios': format_token_as_beios, 'iob': format_token_as_iob}
+    res = []
+    markup = parse_bsf(bsf_markup)
+
+    prev_idx = 0
+    m_ln: BsfInfo
+    for m_ln in markup:
+        res += join_simple_chunk(data[prev_idx:m_ln.start_idx])
+
+        convert_f = converters[converter]
+        res.extend(convert_f(m_ln.token, m_ln.tag))
+        prev_idx = m_ln.end_idx
+
+    if prev_idx < len(data) - 1:
+        res += join_simple_chunk(data[prev_idx:])
+
+    return '\n'.join(res)
+
+
+def parse_bsf(bsf_data: str) -> list:
+    """
+    Convert multiline textual bsf representation to a list of named entities.
+
+    :param bsf_data: data in the format 'T9	PERS 778 783    токен'. Can be multiple lines.
+    :return: list of named tuples for each line of the data representing a single named entity token
+    """
+    if len(bsf_data.strip()) == 0:
+        return []
+
+    ln_ptrn = re.compile(r'(T\d+)\s(\w+)\s(\d+)\s(\d+)\s(.+?)(?=T\d+\s\w+\s\d+\s\d+|$)', flags=re.DOTALL)
+    result = []
+    for m in ln_ptrn.finditer(bsf_data.strip()):
+        bsf = BsfInfo(m.group(1), m.group(2), int(m.group(3)), int(m.group(4)), m.group(5).strip())
+        result.append(bsf)
+    return result
+
+
+def convert_bsf_in_folder(src_dir_path: str, dst_dir_path: str, converter: str = 'beios',
+                          doc_delim: str = '\n', train_test_split_file: str = None) -> None:
+    """
+
+    :param doc_delim: delimiter to be used between documents
+    :param src_dir_path: path to directory with BSF marked files
+    :param dst_dir_path: where to save output data
+    :param converter: `beios` or `iob` output formats
+    :param train_test_split_file: path to file containing train/test lists of file names
+    :return:
+    """
+    # following 2 constants need to comply with stanza naming for corpus and language
+    corpus_name = 'Ukrainian-languk'
+
+    ann_path = os.path.join(src_dir_path, '*.tok.ann')
+    ann_files = glob.glob(ann_path)
+    ann_files.sort()
+
+    tok_path = os.path.join(src_dir_path, '*.tok.txt')
+    tok_files = glob.glob(tok_path)
+    tok_files.sort()
+
+    corpus_folder = os.path.join(dst_dir_path, corpus_name)
+    if not os.path.exists(corpus_folder):
+        os.makedirs(corpus_folder)
+
+    if len(ann_files) == 0 or len(tok_files) == 0:
+        log.warning(f'Token and annotation files are not found at specified path {ann_path}')
+        return
+    if len(ann_files) != len(tok_files):
```

Comment:
What is the consequence of that? Should we just stop?

Reply:
That's a health check. Added it just in case there are some misalignments in data set. So we can at least see that something went wrong. It won't break the logic but the zip() function in further for loop will just iterate on the shortest list.

After looking at it once more - if such a situation happens - the data set will probably be a mess as there is a high chance that .tok and .ann will be misaligned.
In such case we will at least see a log.
Changing it to error level and actually adding return - no point in just printing the message.